expression/builtin_cast.go

```diff
@@ -1687,6 +1687,23 @@ func (i inCastContext) String() string {
 // @see BuildCastFunction4Union
 const inUnionCastContext inCastContext = 0
 
+// hasSpecialCast checks if this expr has its own special cast function.
+// for example(#9713): when doing arithmetic using results of function DayName,
+// "Monday" should be regarded as 0, "Tuesday" should be regarded as 1 and so on.
+func hasSpecialCast(ctx sessionctx.Context, expr Expression, tp *types.FieldType) bool {
+	switch f := expr.(type) {
+	case *ScalarFunction:
+		switch f.FuncName.L {
+		case ast.DayName:
+			switch tp.EvalType() {
+			case types.ETInt, types.ETReal:
```

Comment:
What about first check the type?

Reply:
Checking function name first may be faster? 
There are only a small amount of functions who have a special cast process but almost all functions need to be cast to `Int` or `Real` in some cases.
So in most cases, checking function name first will do `switch-case` only once while it have to do it twice if we check type first.